cvmfs/download.cc

```diff
@@ -1879,15 +1900,26 @@ bool DownloadManager::ProbeHostsGeo() {
   }
   if (!success) {
     LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
-             "failed to retrieve geographic order of stratum 1 servers");
+             "failed to retrieve geographic order from stratum 1 servers");
     return false;
   }
 
   pthread_mutex_lock(lock_options_);
-  delete opt_host_chain_;
+  // sort host chain and fallback proxies by the geo order
+  vector< vector<ProxyInfo> > tmp_fallback_groups((*opt_proxy_groups_).begin() + opt_proxy_groups_fallback_,
+                                                  (*opt_proxy_groups_).end());
+  unsigned hosti = 0;
+  unsigned proxyi = opt_proxy_groups_fallback_;
+  for (unsigned i = 0; i < geo_order.size(); ++i) {
+    uint64_t orderval = geo_order[i];
+    if (orderval < (uint64_t) first_geo_fallback)
+      (*opt_host_chain_)[hosti++] = host_chain[orderval];
+    else
+      (*opt_proxy_groups_)[proxyi++] = tmp_fallback_groups[orderval - first_geo_fallback];
```

Comment:
I just realized that it isn't safe to simply overwrite the indexes in opt_host_chain_ and opt_proxy_groups_ because their lengths may have been changed by another thread in the meanwhile.  I need to recreate them the way that opt_host_chain_ was re-created before I changed it.